fairmq/zeromq/FairMQMessageZMQ.h

```diff
@@ -31,26 +32,33 @@ class FairMQMessageZMQ : public FairMQMessage
     FairMQMessageZMQ(void* data, const size_t size, fairmq_free_fn* ffn, void* hint = nullptr);
     FairMQMessageZMQ(FairMQUnmanagedRegionPtr& region, void* data, const size_t size);
 
-    virtual void Rebuild();
-    virtual void Rebuild(const size_t size);
-    virtual void Rebuild(void* data, const size_t size, fairmq_free_fn* ffn, void* hint = nullptr);
+    void Rebuild() override;
+    void Rebuild(const size_t size) override;
+    void Rebuild(void* data, const size_t size, fairmq_free_fn* ffn, void* hint = nullptr) override;
 
-    virtual void* GetMessage();
-    virtual void* GetData();
-    virtual size_t GetSize();
+    void* GetMessage() override;
+    void* GetData() override;
+    size_t GetSize() const override;
 
-    virtual void SetMessage(void* data, const size_t size);
+    bool SetUsedSize(const size_t size) override;
+    void ApplyUsedSize();
 
-    virtual FairMQ::Transport GetType() const;
+    void SetMessage(void* data, const size_t size) override;
 
-    virtual void Copy(const std::unique_ptr<FairMQMessage>& msg);
+
+    FairMQ::Transport GetType() const override;
+
+    void Copy(const std::unique_ptr<FairMQMessage>& msg) override;
 
     void CloseMessage();
 
-    virtual ~FairMQMessageZMQ();
+    ~FairMQMessageZMQ() override;
 
   private:
-    zmq_msg_t fMessage;
+    bool fUsedSizeModified;
+    size_t fUsedSize;
+    std::unique_ptr<zmq_msg_t> fMsg;
+    std::unique_ptr<zmq_msg_t> fViewMsg; // view on a subset of fMsg (treating it as user buffer)
```

Comment:
Do we need to store `fViewMsg`? Could we just create it on the fly when sending?

Reply:
Sending could still fail/timeout and we should keep the view around for retry/cleanup, even though the size cannot be changed anymore after first send.